### PR TITLE
Lambda person

### PR DESCRIPTION
**Pull request by evgeniyamelone**
### Lambda Expressions And Stream API Practice #1

Create a collection of Person objects (use Person class from the Collections practice). Write a code using Stream API which returns a list of persons names with the following criteria:

- names should be selected only for persons with age between 20 and 30 years; 
- all names should be unique; 
- names in the list should be converted to upper-case; 
- list of names should be sorted by the name length.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>${assertj-core.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jetbrains</groupId>
```
Comment (mih-kopylov) [inline pom.xml:41]: afai remember, this one is not used, isn't it?

File: src/main/java/lambdaperson/FilteredPersons.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import personlist.Person;
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class FilteredPersons {
+
+    public static Set<Person> filterPersons(Set<Person> personList) {
```
Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilteredPersons.java]: A better way is to define a Service class, like `FilterPersonsService` which will have an instance method inside. 
Try to avoid writing static utility methods - they make code organizing hard and usually break OOP principles

File: src/main/java/lambdaperson/FilteredPersons.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        Comparator<Person> sortingByNameLength =
+                Comparator.comparingInt((Person p) -> p.getName().length());
+
+        return personList.stream().filter(Person -> Person.getAge() < 30 & Person.getAge() > 20)
```
Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilteredPersons.java]: `Person` - it's a variable inside lambda, and Java code style recommends to name variables in camelCase, so it's better to name it `person`.
BTW, sometimes it's convenient to give such variables quite short names, such as: `o`, `it`, etc. It's up to author to choose a better one for readability

Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilteredPersons.java]: It's not a good practice to use `&` operator for such cases. `&&` is better.
https://stackoverflow.com/questions/5564410/what-is-the-difference-between-and-in-java

File: src/main/java/lambdaperson/FilteredPersons.java (hunk 2 above)
Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilteredPersons.java]: Should it return just names, according to the task?
 `which returns a list of persons names`

File: src/main/java/lambdaperson/FilteredPersons.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+                .sorted(sortingByNameLength)
+                .map((Person p) -> {
+                    p.setName(p.getName().toUpperCase());
+                    return p;
+                })
+                .collect(Collectors.toCollection(LinkedHashSet::new));
```
Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilteredPersons.java]: So we need to return a sorted collection of smth.
Ther're two main approaches to do that:
1. Return a List that is preliminary sorted
2. Return a Set with TreeSet instantiated with required comparator

Returning a `LinkedHashSet` is technically correct, but seems a bit odd for a general case. Such approach might be used though for some specific case...

File: src/test/java/lambdaperson/FilteredPersonsTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                new Person("John", 25),
+                new Person("Rose", 24),
+                new Person("Stephen", 16),
+                new Person("Stephen", 10),
+                new Person("John", 30)
+        );
+        Set<Person> expected = Set.of(
+                new Person("JOHN", 25),
+                new Person("ROSE", 24),
+                new Person("JOHN", 30),
```
Comment (mih-kopylov) [inline src/test/java/lambdaperson/FilteredPersonsTest.java]: I'm not sure if this test succeeds since there's a condition: `Person.getAge() < 30`
Some tests for boundary values might be helpful

File: src/test/java/lambdaperson/FilteredPersonsTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                new Person("JOHN", 25),
+                new Person("STEPHEN", 26));
+        assertThat(ListActions.equalSets(expected, FilteredPersons.filterPersons(given))).isTrue();
+    }
+
+    @Test
+    void filteredWithEmptySet() {
+        Set<Person> given = Collections.EMPTY_SET;
+
+        Set<Person> expected = Set.of(
```
Comment (mih-kopylov) [inline src/test/java/lambdaperson/FilteredPersonsTest.java]: `expected` is not used. Copy-paste driven development :D

**Review by mih-kopylov — CHANGES_REQUESTED** [review]

File: src/main/java/lambdaperson/FilterPersonsService.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class FilterPersonsService {
+
+    public List<String> filter(Set<Person> personList) {
+        return personList.stream().filter(person -> person.getAge() < 30 && person.getAge() > 20)
+                .peek((Person p) -> p.setName(p.getName().toUpperCase()))
```
Comment (mih-kopylov) [inline src/main/java/lambdaperson/FilterPersonsService.java:13]: It seems I've missed that, sorry.

It's a bad practice to mutate a stream element in stream processing. Much better way is to create a Person with adjusted names.
But in the next step there's `map(Person::getName)`, so that means that there's no need to change `Person#name` at all: after `filter` one can make `map(Person::getName)`, then map the resulting `name` with `toUpperCase` and then sort then by lenght.
